### ISSUE TITLE
[1.13.1] Evaluating `scmVersion.version` prevents further updates to the extension 

### ISSUE THREAD

**Issue by mateuszkwiecinski** [allegro/axion-release-plugin#392]
### Context
Using Gradle extensions often requies caling `afterEvaluate` block to read user data as they are not available right after registering an extension.
I have to defer configuring `axion-release-plugin` exactly for that reason. After migrating to 1.13.1 I noticed there is an additional limiatation introduced in that version. 

### Repro steps
Given following code and having a tag `prefix1.2.3` on one of previous commits
```
plugins {
    id 'pl.allegro.tech.build.axion-release' version '1.13.1'
}

// logger.quiet("Invalid version $scmVersion.version")

afterEvaluate {
    scmVersion.tag {
        prefix = "prefix"
    }

    logger.quiet("Valid version $scmVersion.version")
    version = scmVersion.version
}
```
`./gradlew currentVersion` will print `1.2.3-SNAPSHOT`, which is expected.

However, If I uncomment the code, the `./gardlew currentVersion` tasks returns `0.1.0-SNAPSHOT`. 

### Expected result
I'd expect evaluating `scmVersion.version` doesn't have side effects.

Unfortunately I couldn't find in the code why it happens and what the consequences might be. That's definitely a behavior change when compared with 1.12.1 and doesn't seem to be expected behavior in Gradle ecosystem.

**Comment by bgalek** [comment]
I'll look at this tomorrow! Thanks for detailed issue submission!

**Comment by bgalek** [comment]
@grational isn't that also Your case?

**Comment by mateuszkwiecinski** [comment]
Just FYI: `afterEvaluate` block isn't needed. Simpler repro could be:
```groovy
logger.quiet("Invalid version $scmVersion.version")
scmVersion.tag.prefix = "prefix"
logger.quiet("Valid version $scmVersion.version")
```
the latter will always print `0.1.0-SNAPSHOT`

**Comment by dominikbrandon** [comment]
Ok, I got it now. Thanks for the explanation. You're right that this gets cached. You can see that [here](https://github.com/allegro/axion-release-plugin/blob/b9a6fb1c430bfd3b68ba2ff2e31198b1055364f0/src/main/groovy/pl/allegro/tech/build/axion/release/domain/VersionConfig.groovy#L160). When you look inside the `ensureVersionExists()`, you'll notice that the version gets resolved only once, at the first attempt. It's unclear to me why it works like that - I can only suspect it's for performance reasons. Maybe @bgalek knows what's behind that?

I can see two approaches here. First is to drop this caching mechanism, which sounds unsafe considering the fact that somebody explicitly made it work like that. We can only do that knowing the author's intention. Second is to introduce another method, e.g. `getUncachedVersion()`, which would return the version omitting the cached one. Actually there is already such a method, but it doesn't work as intended, because there is still the context cached.

@mateuszkwiecinski would such another accessor solve the problem for you?
@bgalek what do you think about that? Which way do you suggest taking?

**Comment by manideepla** [comment]
HI, sorry for bringing up an old issue, but I'm getting this while running a CI pipeline. The ```./gradlew currentVersion``` is always returning 0.1.0 no matter what the latest tag is.

I'm depending on this output to do something else. Can I get some hint on overcoming this please?

**Comment by bgalek** [comment]
hi! 
maybe you did a sparse checkout/shallow clone? 
check if you have full repository history since it's needed to get current version ;)

**Comment by manideepla** [comment]
No, it's a clean checkout via the the typical Azure Devops pipeline task. Strangely, it's working okay locally though.

**Comment by dominikbrandon** [comment]
Maybe Azure isn't fetching the tags?

**Comment by manideepla** [comment]
It actually is, it's working in a different pipeline well.
